### DISCUSSION TITLE
Docker installation

### DISCUSSION THREAD

**Discussion by mariushosting** [ananthakumaran/paisa#48] [Ideas]
Is there any Docker image available?

**Comment by Sarath234** [comment]
I have created a simple Dockerfile for Docker Installation 

```
FROM --platform=linux/amd64 ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt update
RUN apt upgrade -y
RUN apt-get install wget -y

RUN wget https://github.com/ananthakumaran/paisa/releases/download/v0.5.2/paisa-cli-linux-amd64
RUN mv paisa-cli-linux-amd64 paisa
RUN chmod u+x paisa
RUN mv paisa /usr/local/bin
EXPOSE 7500
CMD ["/usr/local/bin/paisa", "serve"]

**Reply by eftakhairul** [reply]
I was thinking about it. Thanks for sharing here. I am going to use it.

**Comment by ananthakumaran** [comment]
I am going to convert this to an discussion, please use issue tracker for bugs only

**Comment by ananthakumaran** [comment]
I do have plans to support docker image, just need to figure out the details on the publishing side.

**Comment by ananthakumaran** [comment]
@mariushosting @eftakhairul I have just published docker images to [ananthakumaran/paisa](https://hub.docker.com/r/ananthakumaran/paisa/tags). You can try it, it's not a tagged release, I will do tagged release from the next version.

**Reply by eftakhairul** [reply]
Great work :)

**Reply by Kk-ships** [reply]
@ananthakumaran How do I ensure that transactions data is persistent over the container shutdown and starting?

**Reply by ananthakumaran** [reply]
```
docker run -p 7500:7500 -v /home/john/Documents/paisa/:/root/Documents/paisa/ ananthakumaran/paisa:ci-docker-setup
```

You can mount the directory to host system, all your data would be then stored on `/home/john/Documents/paisa`.

There are a few corner cases with how file names are handled, which I have fixed on master (will be included in the next release). Let me know if you get any error

**Reply by jostyee** [reply]
the Docker image refuses to launch on my Synology DS918+(CPU architecture: x86 64bit):

```
Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "./paisa": stat ./paisa: permission denied: unknown
```

**Reply by ananthakumaran** [reply]
strange, I can run without any issue (x86_64)

```
docker run --rm -p 7500:7500 ananthakumaran/paisa:latest
INFO    Generating config file: /root/Documents/paisa/paisa.yaml
INFO    Generating journal file: /root/Documents/paisa/main.ledger
INFO    Using config file: /root/Documents/paisa/paisa.yaml
INFO    Listening on http://localhost:7500
```

could you share the full command you used?

**Reply by jostyee** [reply]
my bad, I used a docker-compose.yml to start the container, w/o noticing that I set a wrong uid. :-(

**Comment by suyash192** [comment]
@ananthakumaran running the hledger docker image with a different user results in permission errors:

```
paisa  | 	/usr/src/paisa/internal/config/config.go:188 +0x2b
paisa  | github.com/ananthakumaran/paisa/internal/config.init()
paisa  | 	/usr/src/paisa/internal/config/config.go:188 +0x2b
paisa  | 	/usr/src/paisa/internal/config/config.go:188 +0x2b
paisa  | panic: stat .: permission denied
paisa  | panic: stat .: permission denied
```

I don't see any reason for paisa to require root permissions.

my docker compose:

```
services:
  paisa:
    image: ananthakumaran/paisa:latest-hledger
    container_name: paisa
    user: 1000:1000
    ports:
      - 7500:7500
    volumes:
      - /redacted/:/root/Documents/paisa/
    restart: 'unless-stopped'
```

**Reply by suyash192** [reply]
Changed the working directory in the repo Dockerfile

```
WORKDIR /app
ENV HOME=/app
```

And also changed some permissions

```
RUN chmod +x /usr/bin/paisa && chmod -R 777 /app
```

And now the image looks to be working fine

Resulting Dockerfile

```
FROM node:18-alpine3.18 AS web
WORKDIR /usr/src/paisa
COPY package.json package-lock.json* ./
RUN npm install
COPY . .
RUN npm run build

FROM golang:1.21-alpine3.18 AS go
WORKDIR /usr/src/paisa
RUN apk --no-cache add sqlite gcc g++
COPY go.mod go.sum ./
RUN go mod download && go mod verify
COPY . .
COPY --from=web /usr/src/paisa/web/static ./web/static
RUN CGO_ENABLED=1 go build

FROM alpine:3.18
RUN apk --no-cache add ca-certificates ledger tzdata

# Set up a working directory that any user can access
WORKDIR /app
ENV HOME=/app
COPY --from=go /usr/src/paisa/paisa /usr/bin/paisa

# Ensure all files have proper permissions for any user
RUN chmod +x /usr/bin/paisa && chmod -R 777 /app

EXPOSE 7500
CMD ["paisa", "serve"]
```

Similar changes might need to be done in Dockerfile.hledger and other variants.
